### PR TITLE
feat!: remove default Schnorr functionality

### LINKED ISSUE #178: Enforce domain separation on Schnorr signatures
Currently, Schnorr signatures allow for an [optional domain separator](https://github.com/tari-project/tari-crypto/blob/2caaac2b2671551550f7f3952dce85f6231936ec/src/signatures/schnorr.rs#L47-L52), and use a default domain separator if one is not provided.

This is risky, and has already led to unintentional [misuse](https://github.com/tari-project/tari/pull/5394).

While using a default domain separator helps with [simplicity](https://github.com/tari-project/tari-crypto/pull/145#discussion_r1015375102), it may be more trouble than it's worth.

I think the implementation should be changed to mandate a specified domain separator, which forces the implementer to at least think about, and hopefully opt for, safe choices.

### PR DESCRIPTION
**Pull request by AaronFeickert**
Removes default Schnorr signature functionality by requiring domain separation. Closes #178.

Previously, the Schnorr signature API permitted the use of a default domain separator. This introduces significant risk, as reuse of signatures across contexts can be dangerous. Further, as there are no widely-used standards for Schnorr-type signatures over the Ristretto group, there is no particularly good choice for such a domain separator.

This PR does several things:
- requires that `SchnorrSignature` be provided with a domain separator via trait bound, which was previously optional
- removes `RistrettoSchnorrWithDomain`
- redefines `RistrettoSchnorr` to `RistrettoSchnorr<H>` to require a domain separator

It also removes all existing WASM and FFI signing and verification functionality for Schnorr signatures, since this previously used the default domain separator and there is no way to make them properly generic. Applications needing Schnorr signatures using WASM or FFI should define specific functions that use the updated types to provide a domain separator.

BREAKING CHANGE: Makes several type changes to `SchnorrSignature`, `RistrettoSchnorr`, and `RistrettoSchnorrWithDomain` that are breaking.

### CODE REVIEW

File: src/ristretto/ristretto_sig.rs [hunk 1]
```diff
@@ -71,202 +60,152 @@ use crate::{
...
-///     "bd0b253a619310340a4fa2de54cdd212eac7d088ee1dc47e305c3f6cbd020908",
-/// )
-/// .unwrap();
 /// # #[allow(non_snake_case)]
-/// let P = RistrettoPublicKey::from_secret_key(&k);
-/// let sig: SchnorrSignature<RistrettoPublicKey, RistrettoSecretKey, MyCustomDomain> =
-///     SchnorrSignature::sign_message(&k, msg).unwrap();
+/// let (k, P) = get_keypair();
+/// let msg = "Small Gods";
+/// let sig: RistrettoSchnorr<ExampleDomain> = SchnorrSignature::sign_message(&k, msg).unwrap();
```
Comment (hansieodendaal) [inline src/ristretto/ristretto_sig.rs:75]:
```suggestion
/// let sig = RistrettoSchnorr::<ExampleDomain>::sign_message(&k, msg).unwrap();
```

File: src/ristretto/ristretto_sig.rs [hunk 2]
```diff
@@ -71,202 +60,152 @@ use crate::{
...
+        ristretto::{RistrettoPublicKey, RistrettoSchnorr, RistrettoSecretKey},
     };
 
+    hash_domain!(TestDomain, "com.example.test");
+
+    /// Test defaults
     #[test]
     fn default() {
-        let sig = RistrettoSchnorr::default();
+        let sig: RistrettoSchnorr<TestDomain> = RistrettoSchnorr::default();
```
Comment (hansieodendaal) [inline src/ristretto/ristretto_sig.rs:98]:
```suggestion
        let sig = RistrettoSchnorr::<TestDomain>::default();
```

File: src/ristretto/ristretto_sig.rs [hunk 3]
```diff
@@ -71,202 +60,152 @@ use crate::{
...
+        // Use raw signing, where we construct a challenge manually (but without domain separation for this example)
         let e = Blake256::new()
             .chain(P.as_bytes())
             .chain(R.as_bytes())
             .chain(b"Small Gods")
             .finalize();
         let e_key = RistrettoSecretKey::from_bytes(&e).unwrap();
         let s = &r + &e_key * &k;
-        let sig = RistrettoSchnorr::sign_raw(&k, r, &e).unwrap();
+        let sig: RistrettoSchnorr<TestDomain> = RistrettoSchnorr::sign_raw(&k, r, &e).unwrap();
```
Comment (hansieodendaal) [inline src/ristretto/ristretto_sig.rs:120]:
```suggestion
        let sig = RistrettoSchnorr::<TestDomain>::sign_raw(&k, r, &e).unwrap();
```

File: src/ristretto/ristretto_sig.rs [hunk 4]
```diff
@@ -71,202 +60,152 @@ use crate::{
...
-    #[allow(non_snake_case)]
-    fn challenge_from_invalid_scalar() {
-        let mut rng = rand::thread_rng();
-        let m = from_hex("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff").unwrap();
-        let k = RistrettoSecretKey::random(&mut rng);
-        let r = RistrettoSecretKey::random(&mut rng);
-        assert!(RistrettoSchnorr::sign_raw(&k, r, &m).is_ok());
+        // Calculate summand signatures using each component
+        let s1: RistrettoSchnorr<TestDomain> = RistrettoSchnorr::sign_raw(&k1, r1, &e).unwrap();
+        let s2: RistrettoSchnorr<TestDomain> = RistrettoSchnorr::sign_raw(&k2, r2, &e).unwrap();
```
Comment (hansieodendaal) [inline src/ristretto/ristretto_sig.rs:159]:
```suggestion
        let s1 = RistrettoSchnorr::<TestDomain>::sign_raw(&k1, r1, &e).unwrap();
        let s2 = RistrettoSchnorr::<TestDomain>::sign_raw(&k2, r2, &e).unwrap();

```

File: src/ristretto/ristretto_sig.rs [hunk 5]
```diff
@@ -71,202 +60,152 @@ use crate::{
...
-        assert!(!sig.verify_message(&P, "Qs are things that happen to other people"));
-        assert!(!sig.verify_message(&(&P + &P), "Queues are things that happen to other people"));
+        let (_, evil_P) = RistrettoPublicKey::random_keypair(&mut rng);
+
+        // Generate messages
+        let msg = "Queues are things that happen to other people";
+        let evil_msg = "Qs are things that happen to other people";
+
+        // Sign a message
+        let sig: RistrettoSchnorr<TestDomain> = RistrettoSchnorr::sign_message(&k, msg).unwrap();
```
Comment (hansieodendaal) [inline src/ristretto/ristretto_sig.rs:204]:
```suggestion
        let sig = RistrettoSchnorr::<TestDomain>::sign_message(&k, msg).unwrap();
```

File: src/ristretto/ristretto_sig.rs [hunk 6]
```diff
@@ -3,66 +3,55 @@
...
 /// ```rust
+/// # use tari_crypto::hash_domain;
 /// # use tari_crypto::ristretto::*;
 /// # use tari_crypto::keys::*;
 /// # use tari_crypto::signatures::SchnorrSignature;
 /// # use tari_crypto::hash::blake2::Blake256;
 /// # use digest::Digest;
 ///
+/// hash_domain!(ExampleDomain, "com.example");
+///
```
Comment (hansieodendaal) [inline src/ristretto/ristretto_sig.rs:35]:
```suggestion
/// type RistrettoSchnorr<H> = SchnorrSignature<RistrettoPublicKey, RistrettoSecretKey, H>;
///
```

File: src/ristretto/ristretto_sig.rs [hunk 7]
```diff
@@ -71,202 +60,152 @@ use crate::{
...
-/// let P = RistrettoPublicKey::from_secret_key(&k);
-/// let sig: SchnorrSignature<RistrettoPublicKey, RistrettoSecretKey> =
-///     SchnorrSignature::sign_message(&k, msg).unwrap();
-/// assert!(sig.verify_message(&P, msg));
-/// ```
-pub type RistrettoSchnorr = SchnorrSignature<RistrettoPublicKey, RistrettoSecretKey, SchnorrSigChallenge>;
-
-/// # A Schnorr signature implementation on Ristretto with a custom domain separation tag
+/// hash_domain!(ExampleDomain, "com.example");
 ///
```
Comment (hansieodendaal) [inline src/ristretto/ristretto_sig.rs:64]:
```suggestion
/// type RistrettoSchnorr<H> = SchnorrSignature<RistrettoPublicKey, RistrettoSecretKey, H>;
///
```

File: src/ristretto/ristretto_sig.rs [hunk 8]
```diff
@@ -3,66 +3,55 @@
...
 ///     let pk = RistrettoPublicKey::from_secret_key(&k);
 ///     (k, pk)
 /// }
 ///
-/// #[allow(non_snake_case)]
+/// # #[allow(non_snake_case)]
 /// let (k, P) = get_keypair();
 /// let msg = "Small Gods";
-/// let sig = RistrettoSchnorr::sign_message(&k, &msg);
+/// let sig: RistrettoSchnorr<ExampleDomain> = RistrettoSchnorr::sign_message(&k, &msg).unwrap();
```
Comment (hansieodendaal) [inline src/ristretto/ristretto_sig.rs:46]:
```suggestion
/// let sig = RistrettoSchnorr::<ExampleDomain>::sign_message(&k, &msg).unwrap();
```

**Review by hansieodendaal — APPROVED** [review]
I think the changes are good, and agree that the domain should be enforced. I just have some nits about the readability in the test examples, shown below.

ACK

**Comment by CjS77** [comment]
I disagree here. As mentioned out in #178, I feel that API ergonomics is important. Newcomers to the library are going to be confused with an API requiring Domain separation and will use a default anyway. Those looking for it will see it in the function signature and will be able to provide their own domains.

**Comment by CjS77** [comment]
A lot of tests have been removed -- is there a reason for the decrease in coverage?

**Review by CjS77 — CHANGES_REQUESTED** [review]
Need consensus before this can be merged.

**Comment by AaronFeickert** [comment]
> I disagree here. As mentioned out in #178, I feel that API ergonomics is important. Newcomers to the library are going to be confused with an API requiring Domain separation and will use a default anyway. Those looking for it will see it in the function signature and will be able to provide their own domains.

I think implementation snafus like [this](https://github.com/tari-project/tari/pull/5400) are indicative of the risk of providing "a Schnorr signature". The implementer is always free to ignore the recommendation of distinct domain separation, but this API change at least forces them to make a decision. Coupled with the fact that there is no "default" signature by any common standard, I think it's an annoyance whose safety benefit outweighs the inconvenience.

I understand where you're coming from, but respectfully disagree with the conclusion. Frankly, an implementer who is confused by domain separation should probably take a moment to investigate first.

**Comment by AaronFeickert** [comment]
> A lot of tests have been removed -- is there a reason for the decrease in coverage?

Some were specific to FFI or WASM, and naturally went away. An "invalid scalar" test didn't really make sense, and was nixed. A couple of other tests dealt specifically with the default separator, and are no longer relevant or were modified accordingly.

**Comment by AaronFeickert** [comment]
One point to keep in mind is that "raw" signatures, in which the caller is responsible for challenge generation, have no inherent notion of domain separation. Currently, it's probably easiest simply to use the default separator type, since it plays no role in generating the Fiat-Shamir challenge. These are used (for good reason) in several places within the `tari` codebase,

**Pull request closed without merging**